File: AlgorithmLcr/Try3-Component types with known topology and corresponding positions/topofit_id/adjacency.py
"""Unified adjacency-matrix output (../../OUTPUT_FORMAT.md).

Places the fitted parameter groups of a known-topology identification into
the canonical upper-triangle adjacency matrix: ``rows[i][j]`` (i < j) is the
vector of all edges directly connecting nodes i and j.  Nodes 0 and 1 are
the one-port terminals; node labels keep their original (input-graph)
numbering, so nodes removed by reduction survive as empty slots.

Edge follows the root spec: (type, parameter, dcr).  Fitting happens on
reduced groups (F1-F4), so the matrix carries **aggregate** values: a
parallel-merged group is one edge on its slot, a series-merged group is one
edge across its outer nodes (intermediate nodes stay empty), an F4-absorbed
resistor lives on in the inductor group's dcr, and F1-dropped edges are
omitted and reported through ``adjacency_notes`` instead.
"""

from __future__ import annotations

from dataclasses import dataclass

from .fit import FitResult


@dataclass(frozen=True)
class Edge:
    """One 2-terminal element; see OUTPUT_FORMAT.md section 1."""

    type: str          # "R" | "L" | "C"
    parameter: float   # R[ohm] / L[H] / C[F]
    dcr: float = 0.0   # series DC resistance of L; 0 unless type == "L"


def _fmt_edge(e: Edge) -> str:
    if e.type == "L" and e.dcr != 0.0:
        return f"L {e.parameter:.3e} dcr {e.dcr:.3e}"
    return f"{e.type} {e.parameter:.3e}"


class Adjacency:
    """Strict upper-triangle adjacency matrix of vector<Edge> (spec sec.2)."""

    def __init__(self, V: int):
        if V < 2:
            raise ValueError("a one-port needs at least the 2 terminal nodes")
        self.V = V
        self.rows = [[[] for _ in range(V - 1 - i)] for i in range(V)]

    def slot(self, i: int, j: int) -> list[Edge]:
        """All edges directly connecting i and j; requires i < j."""
        if not (0 <= i < j < self.V):
            raise ValueError(f"slot ({i},{j}) outside upper triangle of V={self.V}")
        return self.rows[i][j - i - 1]

    def add(self, i: int, j: int, edge: Edge) -> None:
        """Append an undirected edge; {i, j} order is normalized."""
        if i == j:
            raise ValueError("self loops are not part of the format")
        if i > j:
            i, j = j, i
        self.slot(i, j).append(edge)

    @property
    def n_edges(self) -> int:
        return sum(len(cell) for row in self.rows for cell in row)

    def occupied(self) -> list[tuple[int, int, list[Edge]]]:
        """Non-empty slots in row-major upper-triangle order (spec sec.4)."""
        out = []
        for i in range(self.V):
            for k, cell in enumerate(self.rows[i]):
                if cell:
                    out.append((i, i + k + 1, cell))
        return out

    def format_block(self, label: int | str | None = None,
                     extra_lines: list[str] | None = None) -> str:
        """Unified print form (spec sec.4); label = candidate rank when given."""
        head = f"adjacency[{label}] " if label is not None else "adjacency "
        lines = [f"{head}V={self.V} (ports 0,1):"]
        for i, j, edges in self.occupied():
            lines.append(f"  ({i},{j}): " + " | ".join(_fmt_edge(e) for e in edges))
        if extra_lines:
            lines.extend("  " + s for s in extra_lines)
        return "\n".join(lines)


def _node_span(res: FitResult) -> int:
    """V = max label + 1 over fitted groups and the original input edges."""
    labels = {n for g in res.groups for n in (g.u, g.v)}
    labels.update(n for (u, v, _k) in res.edges for n in (u, v))
    return max(labels) + 1


def fitresult_to_adjacency(res: FitResult) -> Adjacency:
    """Adjacency matrix of a fitted topology: one aggregate Edge per group,
    placed on the group's original node labels."""
    adj = Adjacency(_node_span(res))
    for g in res.groups:
        dcr = g.value[2] if g.kind == "L" else 0.0
        adj.add(g.u, g.v, Edge(g.kind, g.value[1], dcr))
    return adj


def adjacency_notes(res: FitResult) -> list[str]:
    """Annotation lines for merged/dropped original edges (spec sec.5.3)."""
    notes = []
    for er in res.edges_out:
        if er.status == "merged":
            notes.append(f"e{er.index} ({er.kind}) merged: {er.note}")
        elif er.status == "dropped":
            notes.append(f"e{er.index} ({er.kind}) dropped: {er.note}")
    return notes

File: AlgorithmLcr/Try3-Component types with known topology and corresponding positions/cppversion/src/adjacency.cpp
#include "adjacency.hpp"

#include <algorithm>
#include <cstdio>
#include <set>
#include <stdexcept>

namespace tf {

namespace {

std::string fmtEdge(const Edge& e) {
    char buf[96];
    if (e.type == 'L' && e.dcr != 0.0) {
        std::snprintf(buf, sizeof(buf), "L %.3e dcr %.3e", e.parameter, e.dcr);
    } else {
        std::snprintf(buf, sizeof(buf), "%c %.3e", e.type, e.parameter);
    }
    return buf;
}

}  // namespace

Adjacency::Adjacency(int V) : V_(V) {
    if (V < 2) throw std::invalid_argument("a one-port needs at least the 2 terminal nodes");
    rows_.resize((size_t)V);
    for (int i = 0; i < V; ++i) rows_[i].resize((size_t)V - 1 - i);
}

std::vector<Edge>& Adjacency::slot(int i, int j) {
    if (!(0 <= i && i < j && j < V_))
        throw std::invalid_argument("slot outside upper triangle");
    return rows_[i][(size_t)(j - i - 1)];
}

const std::vector<Edge>& Adjacency::slot(int i, int j) const {
    return const_cast<Adjacency*>(this)->slot(i, j);
}

void Adjacency::add(int i, int j, Edge edge) {
    if (i == j) throw std::invalid_argument("self loops are not part of the format");
    if (i > j) std::swap(i, j);
    slot(i, j).push_back(edge);
}

int Adjacency::nEdges() const {
    int n = 0;
    for (const auto& row : rows_)
        for (const auto& cell : row) n += (int)cell.size();
    return n;
}

std::vector<std::tuple<int, int, const std::vector<Edge>*>> Adjacency::occupied() const {
    std::vector<std::tuple<int, int, const std::vector<Edge>*>> out;
    for (int i = 0; i < V_; ++i)
        for (int k = 0; k < V_ - 1 - i; ++k)
            if (!rows_[i][(size_t)k].empty())
                out.push_back({i, i + k + 1, &rows_[i][(size_t)k]});
    return out;
}

std::string Adjacency::formatBlock(const std::string& label,
                                   const std::vector<std::string>& extraLines) const {
    std::string out;
    out += label.empty() ? "adjacency " : ("adjacency[" + label + "] ");
    out += "V=" + std::to_string(V_) + " (ports 0,1):\n";
    for (const auto& [i, j, edges] : occupied()) {
        char head[32];
        std::snprintf(head, sizeof(head), "  (%d,%d): ", i, j);
        out += head;
        for (size_t q = 0; q < edges->size(); ++q) {
            if (q) out += " | ";
            out += fmtEdge((*edges)[q]);
        }
        out += "\n";
    }
    for (const auto& s : extraLines) out += "  " + s + "\n";
    if (!out.empty()) out.pop_back();
    return out;
}

int nodeSpan(const FitResult& res) {
    std::set<int> labels;
    for (const auto& g : res.groups) {
        labels.insert(g.u);
        labels.insert(g.v);
    }
    for (const auto& [u, v, k] : res.edges) {
        (void)k;
        labels.insert(u);
        labels.insert(v);
    }
    return *labels.rbegin() + 1;
}

Adjacency fitresultToAdjacency(const FitResult& res) {
    Adjacency adj(nodeSpan(res));
    for (const auto& g : res.groups) {
        double dcr = g.kind == 'L' ? g.value.v2 : 0.0;
        adj.add(g.u, g.v, Edge{g.kind, g.value.v1, dcr});
    }
    return adj;
}

std::vector<std::string> adjacencyNotes(const FitResult& res) {
    std::vector<std::string> notes;
    for (const auto& er : res.edgesOut) {
        if (er.status == "merged")
            notes.push_back("e" + std::to_string(er.index) + " (" + std::string(1, er.kind) +
                            ") merged: " + er.note);
        else if (er.status == "dropped")
            notes.push_back("e" + std::to_string(er.index) + " (" + std::string(1, er.kind) +
                            ") dropped: " + er.note);
    }
    return notes;
}

}  // namespace tf
